types/src/transaction/auction_transaction.rs

```diff
@@ -0,0 +1,269 @@
+            }
+            AuctionTransaction::WithdrawBid(args) => {
+                WITHDRAW_BID_TAG.write_bytes(writer)?;
+                args.write_bytes(writer)
+            }
+            AuctionTransaction::Delegate(args) => {
+                DELEGATE_TAG.write_bytes(writer)?;
+                args.write_bytes(writer)
+            }
+            AuctionTransaction::Undelegate(args) => {
+                UNDELEGATE_TAG.write_bytes(writer)?;
+                args.write_bytes(writer)
+            }
+            AuctionTransaction::Redelegate(args) => {
+                REDELEGATE_TAG.write_bytes(writer)?;
+                args.write_bytes(writer)
+            }
+            AuctionTransaction::GetEraValidators(args) => {
+                GET_ERA_VALIDATORS_TAG.write_bytes(writer)?;
+                args.write_bytes(writer)
+            }
+            AuctionTransaction::ReadEraId(args) => {
+                READ_ERA_ID_TAG.write_bytes(writer)?;
+                args.write_bytes(writer)
+            }
+        }
+    }
+
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut buffer = bytesrepr::allocate_buffer(self)?;
+        self.write_bytes(&mut buffer)?;
+        Ok(buffer)
+    }
+
+    fn serialized_length(&self) -> usize {
+        U8_SERIALIZED_LENGTH
+            + match self {
+                AuctionTransaction::AddBid(args)
+                | AuctionTransaction::WithdrawBid(args)
+                | AuctionTransaction::Delegate(args)
+                | AuctionTransaction::Undelegate(args)
+                | AuctionTransaction::Redelegate(args)
+                | AuctionTransaction::GetEraValidators(args)
+                | AuctionTransaction::ReadEraId(args) => args.serialized_length(),
+            }
+    }
+}
+
+impl FromBytes for AuctionTransaction {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (tag, remainder) = u8::from_bytes(bytes)?;
+        let (args, remainder) = RuntimeArgs::from_bytes(remainder)?;
+        match tag {
+            ADD_BID_TAG => Ok((AuctionTransaction::AddBid(args), remainder)),
+            WITHDRAW_BID_TAG => Ok((AuctionTransaction::WithdrawBid(args), remainder)),
+            DELEGATE_TAG => Ok((AuctionTransaction::Delegate(args), remainder)),
+            UNDELEGATE_TAG => Ok((AuctionTransaction::Undelegate(args), remainder)),
+            REDELEGATE_TAG => Ok((AuctionTransaction::Redelegate(args), remainder)),
+            GET_ERA_VALIDATORS_TAG => Ok((AuctionTransaction::GetEraValidators(args), remainder)),
+            READ_ERA_ID_TAG => Ok((AuctionTransaction::ReadEraId(args), remainder)),
+            _ => Err(bytesrepr::Error::Formatting),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn bytesrepr_roundtrip() {
+        let rng = &mut TestRng::new();
+        for _ in 0..10 {
+            bytesrepr::test_serialization_roundtrip(&AuctionTransaction::random(rng));
```

Comment:
Some types are doing randomization, while others (older types?) are doing proptest - is there possibility we could stick to one convention for 2.0? (personally not a big fan of randomization) If we were to go with proptest I'd consider `proptest_derive` macro

Reply:
The older types using proptest predate the `TestRng` being available in `casper-types`.

I strongly prefer the `TestRng` approach and would hope that we could eventually migrate many of the proptests to this style in the future.  It's probably unreasonable to expect that to happen (or exclusively using proptest rather than `TestRng`) for 2.0, but I think we should stick with `TestRng` for all new types being created unless there's a compelling reason to use proptest for a given type.